### DISCUSSION TITLE
Build error: undefined reference to std::regex_iterator

### DISCUSSION THREAD

**Discussion by davidzxs123344** [hahnlab/CAFE5#144] [Q&A] [answered]
I have tried ubuntu and centos7 is happened on both system .code ：
g++ -std=c++11 -O3 -fopenmp -o bin/cafe5 -I. obj/user_data.o obj/simulator.o obj/optimizer.o obj/likelihood_ratio.o obj/cafexp.o obj/probability.o obj/gene_family.o obj/newick_ape_loader.o obj/gamma.o obj/gamma_core.o obj/base_model.o obj/error_model.o obj/matrix_cache.o obj/core.o obj/lambda.o obj/io.o obj/root_equilibrium_distribution.o obj/report.o obj/clade.o obj/optimizer_scorer.o obj/poisson.o obj/execute.o obj/gene_family_reconstructor.o obj/easylogging++.o obj/main.o -lm
obj/clade.o：在函数‘parse_newick(std::string, bool)’中：
clade.cpp:(.text+0x3797)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::regex_iterator(__gnu_cxx::__normal_iterator<char const*, std::string>, __gnu_cxx::__normal_iterator<char const*, std::string>, std::basic_regex<char, std::regex_traits<char> > const&, std::bitset<11ul>)’未定义的引用
clade.cpp:(.text+0x37a4)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::regex_iterator()’未定义的引用
clade.cpp:(.text+0x3839)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator!=(std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> > const&)’未定义的引用
clade.cpp:(.text+0x384e)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()’未定义的引用
clade.cpp:(.text+0x39a3)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator++(int)’未定义的引用
clade.cpp:(.text+0x39e9)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()’未定义的引用
clade.cpp:(.text+0x3ac9)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()’未定义的引用
clade.cpp:(.text+0x3bd9)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()’未定义的引用
clade.cpp:(.text+0x3c24)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()’未定义的引用
clade.cpp:(.text+0x3c81)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()’未定义的引用
obj/clade.o:clade.cpp:(.text+0x3df9): 跟着更多未定义的参考到 std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()
collect2: 错误：ld 返回 1
make: *** [bin/cafe5] 错误 1

**Comment by benfulton** [comment, marked as answer]
I think that translates to "undefined reference to std::regex_iterator". We see this most often when using a version of GCC prior to 4.9, which is the minimum required version. Check to make sure your version of GCC is 4.9 or later.

**Reply by davidzxs123344** [reply]
Yes，and soon after I chenge the gcc version , I got another problem using g++ 13.2.0  \g++ 9\g++7 code:
(base) [root@localhost CAFE5]# sudo make
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/user_data.cpp -o obj/user_data.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/simulator.cpp -o obj/simulator.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/optimizer.cpp -o obj/optimizer.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/likelihood_ratio.cpp -o obj/likelihood_ratio.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/cafexp.cpp -o obj/cafexp.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/probability.cpp -o obj/probability.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/gene_family.cpp -o obj/gene_family.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/newick_ape_loader.cpp -o obj/newick_ape_loader.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/gamma.cpp -o obj/gamma.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/gamma_core.cpp -o obj/gamma_core.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/base_model.cpp -o obj/base_model.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/error_model.cpp -o obj/error_model.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/matrix_cache.cpp -o obj/matrix_cache.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/core.cpp -o obj/core.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/lambda.cpp -o obj/lambda.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/io.cpp -o obj/io.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/root_equilibrium_distribution.cpp -o obj/root_equilibrium_distribution.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/report.cpp -o obj/report.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/clade.cpp -o obj/clade.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/optimizer_scorer.cpp -o obj/optimizer_scorer.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/poisson.cpp -o obj/poisson.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/execute.cpp -o obj/execute.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/gene_family_reconstructor.cpp -o obj/gene_family_reconstructor.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp  -c src/easylogging++.cpp -o obj/easylogging++.o
g++ -std=c++11 -I. -O3 -include config.h -DDOCTEST_CONFIG_DISABLE -fopenmp -c main.cpp -o obj/main.o
g++ -std=c++11 -O3 -fopenmp -o bin/cafe5 -I. obj/user_data.o obj/simulator.o obj/optimizer.o obj/likelihood_ratio.o obj/cafexp.o obj/probability.o obj/gene_family.o obj/newick_ape_loader.o obj/gamma.o obj/gamma_core.o obj/base_model.o obj/error_model.o obj/matrix_cache.o obj/core.o obj/lambda.o obj/io.o obj/root_equilibrium_distribution.o obj/report.o obj/clade.o obj/optimizer_scorer.o obj/poisson.o obj/execute.o obj/gene_family_reconstructor.o obj/easylogging++.o obj/main.o -lm

**Reply by benfulton** [reply]
I don't see any error there?

**Reply by davidzxs123344** [reply]
You mean this code means that I complete install CAFE5?

**Reply by benfulton** [reply]
It looks that way. It should have created bin/cafe5, so you can run that directly, add it to your path, or copy it to a directory like /usr/local/bin to run it more conveniently.

**Reply by davidzxs123344** [reply]
Well thankyou very much for answing.Finally solving my problem!

**Comment by lyn320** [comment]
Hi,
I got the same trouble here,:
clade.cpp:(.text+0x3bd9)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()’未定义的引用
clade.cpp:(.text+0x3c24)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()’未定义的引用
clade.cpp:(.text+0x3c81)：对‘std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()’未定义的引用
obj/clade.o:clade.cpp:(.text+0x3df9): more undefined references to `std::regex_iterator<__gnu_cxx::__normal_iterator<char const*, std::string>, char, std::regex_traits<char> >::operator->()' follow
collect2: error: ld returned 1 exit status
make: *** [bin/cafe5] 错误 1

and I checked my gcc version, it shows:

[banliping@master CAFE5-master]$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/apps/local/software/program/gcc-11.1.0/bin/../libexec/gcc/x86_64-pc-linux-gnu/11.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: ../gcc-11.1.0-source/gcc-11.1.0/configure --enable-languages=c,c++,fortran --enable-checking=release --disable-libstdcxx-pch --enable-libgomp --enable-lto --enable-gold --with-plugin-ld=gold --prefix=/usr/local/gcc-11.1.0
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 11.1.0 (GCC),,which is higher than 4.9 ,it still come out error 1,i don't know how to do next, could you please help me to fix this problem? thanks!

**Comment by benfulton** [comment]
`make clean` followed by `make` will probably fix it, or deleting your whole directory and starting over. I suspect there are still some object files that have been compiled by an earlier version of GCC.
